ArchiSteamFarm/Bot.cs

```diff
@@ -2836,6 +2839,295 @@ public sealed class Bot : IAsyncDisposable {
+
+				if (itemsToTakePerInventorySet.Values.All(value => value.Sets == 0)) {
+					return;
+				}
+
+				HashSet<Steam.Asset> result = GetItemsForFullBadge(inventory, itemsToTakePerInventorySet);
+
+				if (result.Count > 0) {
+					await Actions.SendInventory(result).ConfigureAwait(false);
+				}
+			} finally {
+				SendCompleteSetsSemaphore.Release();
+			}
+		}
+
+		private async Task<Dictionary<uint, byte>?> LoadCardsPerSet(ISet<uint> appIDs) {
+			if ((appIDs == null) || (appIDs.Count == 0)) {
+				throw new ArgumentNullException(nameof(appIDs));
+			}
+
+			switch (ASF.GlobalConfig?.OptimizationMode) {
+				case GlobalConfig.EOptimizationMode.MinMemoryUsage:
+					Dictionary<uint, byte> result = new Dictionary<uint, byte>(appIDs.Count);
+
+					foreach (uint appID in appIDs) {
+						byte cardCount = await ArchiWebHandler.GetCardCountForGame(appID).ConfigureAwait(false);
+
+						if (cardCount == 0) {
+							return null;
+						}
+
+						result.Add(appID, cardCount);
+					}
+
+					return result;
+				default:
+					IEnumerable<Task<(uint AppID, byte Cards)>> tasks = appIDs.Select(async appID => (AppID: appID, Cards: await ArchiWebHandler.GetCardCountForGame(appID).ConfigureAwait(false)));
+					IList<(uint AppID, byte Cards)> results = await Utilities.InParallel(tasks).ConfigureAwait(false);
+
+					if (results.Any(tuple => tuple.Cards == 0)) {
+						return null;
+					}
+
+					return results.ToDictionary(res => res.AppID, res => res.Cards);
+			}
+		}
+
+		internal static HashSet<Steam.Asset> GetItemsForFullBadge(IReadOnlyCollection<Steam.Asset> inventory, IReadOnlyDictionary<(uint RealAppID, Steam.Asset.EType Type, Steam.Asset.ERarity Rarity), (uint SetsToExtract, byte CardsPerSet)> amountsToExtract) {
+			if ((inventory == null) || (inventory.Count == 0) || (amountsToExtract == null) || (amountsToExtract.Count == 0)) {
+				throw new ArgumentNullException(nameof(inventory) + " || " + nameof(amountsToExtract));
+			}
+
+			HashSet<Steam.Asset> result = new HashSet<Steam.Asset>();
+			Dictionary<(uint RealAppID, Steam.Asset.EType Type, Steam.Asset.ERarity Rarity), Dictionary<ulong, List<Steam.Asset>>> itemsPerClassIDPerSet = inventory.GroupBy(item => (item.RealAppID, item.Type, item.Rarity)).ToDictionary(grouping => grouping.Key, grouping => grouping.GroupBy(item => item.ClassID).ToDictionary(group => group.Key, group => group.ToList()));
+
+			foreach (((uint RealAppID, Steam.Asset.EType Type, Steam.Asset.ERarity Rarity) set, (uint setsToExtract, byte cardsPerSet)) in amountsToExtract) {
+				if (!itemsPerClassIDPerSet.TryGetValue(set, out Dictionary<ulong, List<Steam.Asset>>? itemsPerClassID)) {
+					continue;
+				}
+
+				if (cardsPerSet < itemsPerClassID.Count) {
+					throw new ArgumentException(nameof(inventory) + " || " + nameof(amountsToExtract));
```

Comment:
That should be `ArgumentOutOfRangeException` with `nameof(inventory) + " && " + nameof(amountsToExtract)`

Reply:
I just want to mention, that this comment comes from a person throwing `ArgumentNullException` everywhere since his nullable-rewrite, because he is too lazy to throw the proper ones.

Reply:
You can correct throwed exceptions in the next PR if you'd like to 🙂 

And it's still wrong message.

Reply:
It's not the wrong message in my opinion. The values that we check against are directly derived from the arguments of which I put the names in `Argument[...]Exception` . It is those arguments that provide wrong information to us and not the logic or local variables of our (pure) function.

Reply:
Fair point, but we don't want 2 different places to throw exactly same exception in a way it's impossible to tell which place got affected. Perhaps we should simply throw `InvalidOperationException` then instead here.

Reply:
It's not the same exception ;) one has `||` in its string and the other one has `&&`

Reply:
> The exception that is thrown when a method call is invalid for the object's current state.

Sounds rational to me.

Reply:
You're right, it's not the same, still, `InvalidOperationException` makes more sense here as we've already done some calculations and stumbled upon something unsupported based on input.

Reply:
> The object's current state
The object would be `this` and the state would be it's properties. As I previously mentioned: This is a pure function depending only on it's parameters. Shall we still switch? I don't think local variables on the stack count towards the state of the object, but I'm open for discussion on this matter.

Reply:
It's fine I guess, the most important part is lack ambiguity - which is achieved here.